### ISSUE TITLE
Use exponential backoff instead of delay

### ISSUE THREAD

**Issue by dgilman** [dgilman/gspread_asyncio#35]
Let us bounce up against Google's rate limiting instead of trying to figure out how long to sleep ourselves.

**Comment by pavlosambur** [comment]
Hi!

Is this still the direction you'd want? Happy to do the work if so.

Two things I found while looking at it:

gspread 6.x ships `BackOffHTTPClient`, and `gspread.authorize()` takes it
directly:

    gspread.authorize(creds, http_client=BackOffHTTPClient)

`ClientManager._authorize` currently calls `gspread.authorize(creds)`, so
wiring it up is small. The catch is that its own docstring says it is "not
production ready yet", so I wouldn't switch to it silently — probably opt-in
first, with `gspread_delay` kept as the default.

On the current default: `gspread_delay = 1.1` paces every call through one
counter, which works out at roughly 54 calls a minute. Sheets meters reads and
writes separately at 60 each, so a workload that alternates them is running at
well under half the quota. That's a separate question from #35, but it's the
reason the delay is felt in practice.

If you'd rather keep the self-pacing approach than move to backoff, that
changes what's worth doing here — happy either way, just don't want to write
the wrong PR.

**Comment by dgilman** [comment]
Backoff after hitting the rate limits is just fine. Being able to take advantage of the two rate limits is great. You also might get a Retry-After from Google, that's worth respecting too, perhaps as the starting duration for your backoff.

I can't comment on the gspread BackOffHTTPClient because I haven't really reviewed it or ever used it but it's probably in better shape than the docs suggest, their project is just defunct.

You could also configure a `urllib3.util.Retry` for the requests Session that gets passed around. However, given that this program is meant to be used in heavily concurrent asyncio programs, I would prefer if it was using asyncio locks or other synchronization tools to block all callers on a single timeout instead of letting each caller trip and sleep on their own 429 timeout.

**Comment by pavlosambur** [comment]
Thanks. The shared-timeout requirement is what rules out both easy options:
`urllib3.util.Retry` and `BackOffHTTPClient` both sleep inside the worker thread
of a single call, so it has to live in the asyncio layer, in `ClientManager`.

One thing worth spelling out about the current code, because it makes the job
bigger than it looks. `call_lock` is held across `delay()`, the call and the
error handlers, so no two callers ever sleep at once — but there is no shared
"we are rate limited until T" state anywhere. Three concurrent callers against a
fake method that 429s once then succeeds, from an instrumented run — `delay()`
and `handle_gspread_error` logging what they actually slept, elapsed time on the
left (condensed):

     0.0s  A: delay() returns — first call, last_call is None
     0.0s  A: request, attempt 1 -> 429, sleeps 1.1s
     1.1s  B: delay() returns — gap since last call already >= 1.1s
     1.1s  B: request, attempt 1 -> 429, sleeps 1.1s
     2.2s  C: delay() returns — gap since last call already >= 1.1s
     2.2s  C: request, attempt 1 -> 429, sleeps 1.1s
     3.3s  A: request, attempt 2 -> ok

B doesn't wait at all: A's sleep already made `delta >= gspread_delay`, so B's
`delay()` is a no-op and B fires straight into a quota known to be exhausted.
Then C. N callers still burn N separate 429s — serially, but nothing is learned
from the first. And since the retry loop goes back through `acquire()`, A's
retry queues up behind callers already waiting and lands only at 3.3s, after
B's and C's doomed attempts — the more waiters, the longer a retry starves.

So the work is a shared deadline — a timestamp or an `asyncio.Event` consulted
inside `delay()` — rather than a sleep held under a lock. That would also give
#6 something concrete to report.

Three decisions I'd rather ask about than guess.

**1. How much of `call_lock` is in scope.** It's held across
`await asyncio.to_thread(method, ...)`, so exactly one API call is ever in
flight. That means separate read/write pacing buys almost nothing on its own —
a read still can't overlap a write, so the ceiling stays one-request-at-a-time
rather than 60+60/min. Narrowing the lock to just the pacing and backoff
decision is a real behaviour change: it currently also serializes
`before_gspread_call`, which subclasses may rely on. In this issue, or separate?

Attached to it: telling a read from a write. `_call` only sees `method` and
`api_call_count`, so today the only option is a table of gspread method names in
the wrapper — brittle, and silently wrong for anything added upstream. The clean
place is where the verb and URL are visible, which is #44, and its "APIs that
make more than one network call" is the same thing `api_call_count` is papering
over. Name table now, or does this wait on #44?

**2. What the retry contract becomes.** `handle_gspread_error(e, method, args,
kwargs)` returns nothing and just sleeps, so it can't know which attempt it is —
you can't write an exponential from that, and seeding from `Retry-After` needs
the same plumbing. Either extend the signature of a documented extension point,
or move the retry state into `_call` and leave the hook advisory. Along with it:
the default currently retries forever, so an unbounded exponential is an
unbounded delay — does a max delay or max attempts appear?

**3. What upgrading looks like.** Does backoff become the default
implementation, or something callers opt into? That also settles what
`gspread_delay` means afterwards — a floor between calls, the seed for the
backoff, or deprecated. This is the one that decides whether anyone's existing
timing changes under them.

On tests, #4 with the `v1.0` label reads as settled, so this is about layout
rather than whether. Worth noting that no PR runs the integration suite
upstream (`ci.yml` is `on: [push]` with no `pull_request` trigger, and a fork PR
wouldn't get `AGE_KEY` anyway), and rate limiting isn't something you can test
against the live API in any case. So what I'd propose, as its own PR before any
of the above:
`tests/unit.py` next to `tests/integration.py`, a `make test-unit` target
running `python -m unittest tests.unit` with no `.env` and no credentials, and a
CI job that runs it without the age/`decrypt-test-files` steps.
`unittest.IsolatedAsyncioTestCase` covers the async side, so no `async_test`
decorator needed. Contents on the lines of the run above, plus a "two callers,
one 429, one shared deadline" case. Happy to put it elsewhere if you'd rather.

One last thing, possibly out of scope for #35 — say if it should be its own
issue:
`_call` raises on `400 <= code <= 499 and code != 429`, so a 403 carrying
`rateLimitExceeded`/`userRateLimitExceeded` is fatal today rather than
retried.

On `Retry-After` I'll just decide instead of asking: use it as the starting
delay when the response carries it, exponential when it doesn't, so nothing
depends on Google sending it.

**Comment by dgilman** [comment]
Can I get your human to review that and write me a response instead of that big ball of slop?

**Comment by pavlosambur** [comment]
You're right, sorry. That was way too long and a bit complicated))

Short version of what I actually found. 
I ran the library locally, three concurrent callers, fake method that returns one 429 then
succeeds, logging what delay() and handle_gspread_error() really
slept:

  0.00s  A: request -> 429, handler sleeps 1.10s
  1.10s  B: delay() no-op, request -> 429, handler sleeps 1.10s
  2.20s  C: delay() no-op, request -> 429, handler sleeps 1.10s
  3.30s  A: request, attempt 2 -> ok

delay() only measures the gap since the last call, and the previous
caller's error sleep already covers that gap. So B fires straight
into a quota that is already known to be exhausted, then C. Nothing
is learned from the first 429. That's why it stopped looking like a
small change to me, and I dumped everything at once instead of
picking one thing. My fault.

And yes, I used AI for the write-up; I'm not a native English speaker, so I 
need to use something that helps me to communicate. 
The findings I'll stand behind, you can reproduce them.

I'd rather go step by step if you're up for it. First step I'd
suggest is a credential-free unit test, which is #4 anyway: fake
method raising a constructed APIError, injected clock so the test
asserts on the delays the code decided on instead of real waiting,
and a case for "three callers, three separate 429s". Needs no
decisions from you, runs in CI without AGE_KEY, and if I've misread
something that's where it shows.

Then backoff on top of that, and I'll bring the design questions
one at a time when there's something to run them against.

Tell me if you'd rather start somewhere else.
